tensorflow/c/kernels.cc

```diff
@@ -222,7 +239,87 @@ void TF_OpKernelContext_Failure(TF_OpKernelContext* ctx, TF_Status* status) {
   cc_ctx->CtxFailure(s);
 }
 
-#define DEFINE_TF_GETATTR(func, c_type, cc_type)                               \
+TF_AttrMetadata TF_OpKernelConstruction_GetAttrMetadata(
```

Comment:
the primary usecase for this functionality is when the attribution is a vector type, we need to use the Metadata to get the length of the vector so that we can allocate a specified size of memory to store the attribution, and call the GetAttr C API to fill that memory.
yes, it is very similar with TF_OperationGetAttrMetadata since both of them are intended to get node attribution.
I think in the TF_OpKernelConstruction, we will always get a non-empty list. And also in the Op C API, when it find that node attribution list is empty, its meta data type will be set as op's type, but OpKernelConstruction C API can only get the node attribution, I didn't find a good way to set a type for it.

Reply:
Makes sense. Yes, what you describe is my primary concern. If an empty attribute list is a valid input, then this can definitely happen (that is, we don't know the type). Therefore, we can't really promise to fill in the type, at least not in this case.

Therefore, it seems better to me to just have a function that returns list_size instead of all of `TF_AttrMetadata`.

Reply:
I think TF_AttrMetadata is needed, only list_size seems not enough. currently TF_AttrMetadata has (is_list, list_size, type, total_size), list_size indicates the vector length(how many elements in the vector), when the attribution is a vector<string>, we also need to get the total_size to allocate the vector.size() * string.size() of memory for the vector<string> in the plugin side. we may also use is_list to check whether the attribution is scalar or vector.

Reply:
Then, we could have a function that returns `list_size` and `total_size`. A kernel should know type of its attributes and whether they are lists or not. Therefore, `is_list` should not be needed (let me know if I am missing something).

I still feel like using `TF_AttrMetadata` might not be the best. If we return `TF_AttrMetadata` then an empty list will always have type specified as `TF_ATTR_STRING` (`TF_ATTR_STRING` corresponds to `0` [here]( https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/tf_attrtype.h;l=24?q=tensorflow%2Fc%2Ftf_attrtype.h)) since [this logic](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/c_api.cc;l=1286) is missing. This would be confusing when the actually attribute list type is, say, `int`. It would have been better if default type of `TF_AttrType` was `TF_ATTR_UNKNOWN`.

Reply:
thanks for the comment. I think is_list and the datatype is useful, it will help plugin authors to write some error handling code. such as if the Attribute is a int list, however, plugin authors may wrongly call the TF_OpKernelConstruction_GetAttrInt32 and will cause unexpected behaviors.

can we still use the TF_Metadata and make the default type as `TF_ATTR_UNKOWN` or `TF_ATTR_PLACEHOLDER`? I saw that was a existing TF_AttrType is `TF_ATTR_PLACEHOLDER`

plugin example code:
```
template <typename T>
Plugin_OpKernelConstruction_GetAttr(ctx, "name", metadata, T& value){
 if (!metadata.is_list)
    switch(metadata.data_type) {
      case:TF_ATTR_INT:
        TF_OpKernelConstruction_GetAttrInt()
    }
  }  else
      {
    swith(metadata.type)
      case TF_ATTR_INT:
      TF_OpKernelConstruction_GetAttrInt32List().
       }
}

```

Reply:
At least in this case `is_list` and `data_type` checks look redundant. This function would generally be implemented per-kernel and a kernel would take either a list or a scalar (not both). Whether it takes the list or not would be defined as a part of its op definition. That's how existing TensorFlow kernels work.

For example, say here we are just getting an attribute by passing in an `std::vector<int32>` variable to be set by `GetAttr`. We don't check beforehand that the type is actually a list if ints:
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/kernels/conv_ops.cc;l=349;drc=669993ebe8534eac877eec61225925cff737eac2

We could add `TF_ATTR_UNKNOWN` but it should ideally be the first item in the enum. Unfortunately, we can't add it as  first item anymore (due to API guarantees). `TF_ATTR_PLACEHOLDER` has a special meaning. So, we could either add `TF_ATTR_UNKNOWN` or use `TF_ATTR_PLACEHOLDER` but since these are not ideal, we should only do that if we have a good usecase for it.

It might be I am missing something and let me know if thats the case and there is a good usecase for having is_list and type.

Reply:
@annarev I have change the TF_MetaData to list_size + total_size, please have a review. Thanks.